File: Lexer/Lexer.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FrostScript
{
    public static class Lexer
    {
        public static IEnumerable<Token> GetTokens(string sourceCode)
        {
            var characters = sourceCode.ToCharArray();

            var line = 1;
            var characterPos = 0;

            for (int i = 0; i < characters.Length; i++)
            {
                var character = characters[i];
                characterPos++;
                switch (character)
                {
                    case '(': yield return new(TokenType.ParentheseOpen, line, characterPos, character.ToString()); break;
                    case ')': yield return new(TokenType.ParentheseClose, line, characterPos, character.ToString()); break;
                    case '{': yield return new(TokenType.BraceOpen, line, characterPos, character.ToString()); break;
                    case '}': yield return new(TokenType.BraceClose, line, characterPos, character.ToString()); break;
                    case ',': yield return new(TokenType.Comma, line, characterPos, character.ToString()); break;
                    case '.': yield return new(TokenType.Dot, line, characterPos, character.ToString()); break;
                    case '+': yield return new(TokenType.Plus, line, characterPos, character.ToString()); break;
                    case ';': yield return new(TokenType.NewLine, line, characterPos, character.ToString()); break;
                    case '*': yield return new(TokenType.Star, line, characterPos, character.ToString()); break;
                    case '|': yield return new(TokenType.Pipe, line, characterPos, character.ToString()); break;
                    case '-':
                        if (characters.Skip(i + 1).Contains('>'))
                        {
                            yield return new(TokenType.Arrow, line, characterPos + 1, character.ToString());
                            i++;
                            characterPos++;
                        }
                        else yield return new(TokenType.Minus, line, characterPos, character.ToString());

                        break;

                    case '!': yield return Match('=') ? new(TokenType.NotEqual, line, characterPos, character.ToString()) : new(TokenType.Not, line, characterPos, character.ToString()); break;
                    case '=': yield return Match('=') ? new(TokenType.Equal, line, characterPos, character.ToString()) : new(TokenType.Assign, line, characterPos, character.ToString()); break;
                    case '<': yield return Match('=') ? new(TokenType.LessOrEqual, line, characterPos, character.ToString()) : new(TokenType.LessThen, line, characterPos, character.ToString()); break;
                    case '>': yield return Match('=') ? new(TokenType.GreaterOrEqual, line, characterPos, character.ToString()) : new(TokenType.GreaterThen, line, characterPos, character.ToString()); break;

                    //string litteral
                    case '"':
                        if (!characters.Skip(i + 1).Contains('"'))
                            Reporter.Report(line, i + 1, $"string literal was not closed");
                        var stringCharacters = characters.Skip(i + 1).TakeWhile(x => x != '"').ToArray();
                        var stringLit = new string(stringCharacters);
                        yield return new Token(TokenType.String, line, characterPos, stringLit, stringLit);

                        i += stringCharacters.Length + 1;
                        break;

                    //numeral litteral
                    case char _ when char.IsDigit(character):
                        var digits = new string(characters.Skip(i).TakeWhile(x => char.IsDigit(x) || x == '.').ToArray());

                        yield return new Token(TokenType.Numeral, line, characterPos, digits, double.Parse(digits));
                        i += digits.Length - 1;

                        //add a multiplication if numeric is followed by a parenthese
                        if (i + 1 < characters.Length && characters[i + 1] == '(')
                            yield return new Token(TokenType.Star, line, character);

                        break;

                    //ids and reserved words
                    case char _ when char.IsLetter(character):
                        var word = new string(characters.Skip(i).TakeWhile(x => char.IsLetterOrDigit(x)).ToArray());
                        yield return word switch
                        {
                            "if" => new Token(TokenType.If, line, characterPos, word),
                            "else" => new Token(TokenType.Else, line, characterPos, word),
                            "when" => new Token(TokenType.When, line, characterPos, word),

                            "print" => new Token(TokenType.Print, line, characterPos, word),

                            "true" => new Token(TokenType.True, line, characterPos, word, true),
                            "false" => new Token(TokenType.False, line, characterPos, word, false),
                            "null" => new Token(TokenType.Null, line, characterPos, word),

                            "for" => new Token(TokenType.For, line, characterPos, word),
                            "while" => new Token(TokenType.While, line, characterPos, word),

                            "var" => new Token(TokenType.Var, line, characterPos, word),
                            "let" => new Token(TokenType.Let, line, characterPos, word),

                            //new id
                            _ => new Token(TokenType.Id, line, i + 1, word)
                        };

                        i += word.Length - 1;
                        break;


                    //comment
                    case '/':
                        if (Match('/'))
                        {
                            //skip to end of comment
                            i += characters.Skip(i).TakeWhile(x => x != '\n').Count();

                            continue;
                        }
                        else
                            yield return new(TokenType.Slash, line, characterPos, character.ToString());
                        break;

                    //ignore white space
                    case ' ':
                    case '\r':
                    case '\t':
                        break;

                    case '\n': line++; characterPos = 0; break;

                    default: Reporter.Report(line, i + 1, $"Charactor {character} not supported"); break;
                }

                bool Match(char expected)
                {
                    if (i + 1 >= characters.Length)
                        return false;

                    if (characters[i + 1] != expected)
                        return false;

                    i++;
                    return true;
                }
            }

            yield return new Token(TokenType.Eof, line, 0);
        }
    }
}

File: FrostScript/Lexer.fs
namespace FrostScript

module Lexer =
    let lex (rawScript : string) =
        let chars = rawScript.ToCharArray() |> Array.toList
        seq { 
            let mutable character = 0
            let mutable line = 1
            let mutable i = -1
           
            while i < chars.Length - 1 do
                character <- character + 1
                i <- i + 1
                match chars.[i] with
                | ')' -> yield { Type = ParentheseClose; Lexeme = ")"; Literal = None; Line = line; Character = character }
                | '{' -> yield { Type = BraceOpen; Lexeme = "{"; Literal = None; Line = line; Character = character }
                | '}' -> yield { Type = BraceClose; Lexeme = "}"; Literal = None; Line = line; Character = character }
                | ',' -> yield { Type = Comma; Lexeme = ","; Literal = None; Line = line; Character = character }
                | '.' -> yield { Type = Period; Lexeme = "."; Literal = None; Line = line; Character = character }
                | '+' -> yield { Type = Operator Plus; Lexeme = "+"; Literal = None; Line = line; Character = character }
                | '*' -> yield { Type = Operator Multiply; Lexeme = "*"; Literal = None; Line = line; Character = character }
                | '/' -> yield { Type = Operator Devide; Lexeme = "/"; Literal = None; Line = line; Character = character }
                | ';' -> yield { Type = SemiColon; Lexeme = ";"; Literal = None; Line = line; Character = character }
                | ':' -> yield { Type = Colon; Lexeme = ":"; Literal = None; Line = line; Character = character }
                | '>' ->
                    yield
                         match chars.[i + 1] with
                            | '=' ->
                                i <- i + 1
                                character <- character + 1
                                { Type = GreaterOrEqual; Lexeme = ">="; Literal = None; Line = line; Character = character }
                            | _ -> 
                                { Type = GreaterThen; Lexeme = ">"; Literal = None; Line = line; Character = character }

                | '<' -> 
                    yield
                        match chars.[i + 1] with
                        | '=' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = LessOrEqual; Lexeme = "<="; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = LessThen; Lexeme = "<"; Literal = None; Line = line; Character = character }
                | '!' -> 
                    yield 
                        match chars.[i + 1] with
                        | '=' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = NotEqual; Lexeme = "!="; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = Not; Lexeme = "!"; Literal = None; Line = line; Character = character }
                            
                | '=' ->
                    yield 
                        match chars.[i + 1] with
                        | '=' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = Equal; Lexeme = "=="; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = Assign; Lexeme = "="; Literal = None; Line = line; Character = character }
                | '|' -> 
                    yield 
                        match chars.[i + 1] with
                        | '>' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = BlockReturn; Lexeme = "|>"; Literal = None; Line = line; Character = character }
                        | '-' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = Pipe; Lexeme = "|-"; Literal = None; Line = line; Character = character }
                        | '.' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = ObjectPipe; Lexeme = "|."; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = BlockOpen; Lexeme = "|"; Literal = None; Line = line; Character = character }
                | '(' -> 
                    yield 
                        match chars.[i + 1] with
                        | ')' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = Void; Lexeme = "()"; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = ParentheseOpen; Lexeme = "("; Literal = None; Line = line; Character = character }

                | '-' -> 
                    yield 
                        match chars.[i + 1] with
                        | '>' ->
                            i <- i + 1
                            character <- character + 1
                            { Type = Arrow; Lexeme = "->"; Literal = None; Line = line; Character = character }
                        | _ -> 
                            { Type = Operator Minus; Lexeme = "-"; Literal = None; Line = line; Character = character }

                //strings
                | '"' ->
                    if chars
                        |> List.skip (i + 1)
                        |> List.contains '"'
                        |> not
                    then
                        yield { Type = LexerError "String never closed"; Lexeme = string chars.[i]; Literal = None; Line = line; Character = character } 
                        i <- chars.Length - 1
                    else
                        let word = 
                            new string (chars 
                            |> List.skip (i + 1) 
                            |> List.takeWhile (fun x -> x <> '"')
                            |> List.toArray)

                        yield {Type = String; Lexeme = word; Literal = Some word; Line = line; Character = character }
                        i <- i + word.Length + 1
                        character <- character + word.Length + 1

                //labels
                | _ when System.Char.IsLetter chars.[i] ->
                    let word = 
                        new string (chars
                        |> List.skip i
                        |> List.takeWhile (fun x -> System.Char.IsLetterOrDigit(x))
                        |> List.toArray)

                    match word with
                    | "true"   -> yield { Type = Bool; Lexeme = word; Literal = Some true; Line = line; Character = character }
                    | "false"  -> yield { Type = Bool; Lexeme = word; Literal = Some false; Line = line; Character = character }
                    | "let"    -> yield { Type = Let; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "var"    -> yield { Type = Var; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "fun"    -> yield { Type = Fun; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "or"     -> yield { Type = Or; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "and"    -> yield { Type = And; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "num"    -> yield { Type = TypeAnnotation (NumberType); Lexeme = word; Literal = None; Line = line; Character = character }
                    | "any"    -> yield { Type = TypeAnnotation (AnyType); Lexeme = word; Literal = None; Line = line; Character = character }
                    | "void"   -> yield { Type = TypeAnnotation (VoidType); Lexeme = word; Literal = None; Line = line; Character = character }
                    | "bool"   -> yield { Type = TypeAnnotation (BoolType); Lexeme = word; Literal = None; Line = line; Character = character }
                    | "string" -> yield { Type = TypeAnnotation (StringType); Lexeme = word; Literal = None; Line = line; Character = character }
                    | "if"     -> yield { Type = If; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "else"   -> yield { Type = Else; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "for"    -> yield { Type = For; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "while"  -> yield { Type = While; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "do"     -> yield { Type = Do; Lexeme = word; Literal = None; Line = line; Character = character }
                    | "new"     -> yield { Type = New; Lexeme = word; Literal = None; Line = line; Character = character }
                    | _        -> yield { Type = Id; Lexeme = word; Literal = None; Line = line; Character = character }

                    i <- i + word.Length - 1
                    character <- character + word.Length

                //numbers
                | _ when System.Char.IsDigit chars.[i] ->
                    let getNumber chars = 
                        chars
                        |> List.takeWhile (fun x -> System.Char.IsDigit x)
                        |> List.toArray

                    let wholeNumber = getNumber (chars |> List.skip i)

                    let number = 
                        if chars.Length > i + wholeNumber.Length && chars.[i + wholeNumber.Length] = '.' then
                            let decimals = getNumber (chars |> List.skip (i + wholeNumber.Length + 1))
                            new string(
                                wholeNumber
                                |> Array.append [|'.'|]
                                |> Array.append decimals
                                |> Array.rev
                            )
                        else
                            new string(wholeNumber)

                    i <- i + number.Length - 1
                    character <- character + number.Length - 1

                    yield { Type = Number; Lexeme = number; Literal = Some (double number); Line = line; Character = character }

                //ignore whitespace
                | ' ' | '\t' -> ()

                //new line
                | '\r' -> 
                    if chars.[i + 1] = '\n' then
                        i <- i + 1
                        line <- line + 1
                        character <- 0
                    else
                        line <- line + 1
                        character <- 0
                        
                | '\n' -> 
                    line <- line + 1
                    character <- 0

                | _ -> yield { Type = LexerError "Unexpected character"; Lexeme = string chars.[i]; Literal = None; Line = line; Character = character } 

        } |> Seq.toList